src/main/java/com/google/googleidentity/oauth2/token/InMemoryOAuth2TokenService.java

```diff
@@ -0,0 +1,457 @@
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    public Optional<OAuth2RefreshToken> readRefreshToken(String refreshToken) {
+        UserClientTokenInfo info = null;
+
+        try{
+            info = new UserClientTokenInfo(refreshToken);
+        } catch (InvalidParameterException exception){
+            log.log(Level.INFO,  "Invalid refresh token Value", exception);
+            return Optional.empty();
+        }
+
+        String username = info.getUsername();
+        String clientID = info.getClientID();
+        String tokenValue = info.getTokenValue();
+
+        if (userClientRefreshTokenMap.containsKey(username) &&
+                userClientRefreshTokenMap.get(username).containsKey(clientID)) {
+            return Optional.ofNullable(
+                    userClientRefreshTokenMap.get(username).get(clientID).get(tokenValue));
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    public boolean revokeAccessToken(String accessToken) {
+        UserClientTokenInfo info = null;
+
+        try{
+            info = new UserClientTokenInfo(accessToken);
+        } catch (InvalidParameterException exception){
+            log.log(Level.INFO,  "Invalid refresh token Value", exception);
+            return false;
+        }
+
+        String username = info.getUsername();
+        String clientID = info.getClientID();
+        String tokenValue = info.getTokenValue();
+
+        if (userClientAccessTokenMap.containsKey(username) &&
+                userClientAccessTokenMap.get(username).containsKey(clientID) &&
+                userClientAccessTokenMap.get(username).get(clientID).containsKey(tokenValue)) {
+            userClientAccessTokenMap.get(username).get(clientID).remove(tokenValue);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    public boolean revokeRefreshToken(String refreshToken) {
+        UserClientTokenInfo info = null;
+
+        try{
+            info = new UserClientTokenInfo(refreshToken);
+        } catch (InvalidParameterException exception){
+            log.log(Level.INFO,  "Invalid refresh token Value", exception);
+            return false;
+        }
+
+        String username = info.getUsername();
+        String clientID = info.getClientID();
+        String tokenValue = info.getTokenValue();
+
+        if (userClientRefreshTokenMap.containsKey(username) &&
+                userClientRefreshTokenMap.get(username).containsKey(clientID) &&
+                userClientRefreshTokenMap.get(username).get(clientID).containsKey(tokenValue)) {
+            userClientRefreshTokenMap.get(username).get(clientID).remove(tokenValue);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    public List<String> listUserClient(String username) {
+        Set<String> clientList = new HashSet<>();
+        if (userClientAccessTokenMap.containsKey(username)) {
```

Comment:
Since the client list in userClientAccessTokenMap and userClientRefreshTokenMap should be the same, I suggest to put a check here and throw exception when they are different.